docs/jax-101/07-state.md

```diff
@@ -0,0 +1,283 @@
+
+
+```{code-cell} ipython3
+:id: S-n8XRC2kTnI
+
+import jax
+import jax.numpy as jnp
+```
+
++++ {"id": "s_-6semKkSzp"}
+
+
+
+## A simple example: Counter
+
+Let's start by looking at a simple stateful program: a counter.
+
+```{code-cell} ipython3
+---
+id: B3aoCHpjg8gm
+outputId: 5cbcfbf5-5c42-498f-a175-050438518337
+---
+class Counter:
+  """A simple counter."""
+
+  def __init__(self):
+    self.n = 0
+
+  def count(self) -> int:
+    """Increments the counter and returns the new value."""
+    self.n += 1
+    return self.n
+
+  def reset(self):
+    """Resets the counter to zero."""
+    self.n = 0
+
+
+counter = Counter()
+
+for _ in range(3):
+  print(counter.count())
+```
+
++++ {"id": "SQ-RNLfdiw04"}
+
+The `n` attribute maintains the counter's _state_ between successive calls of `count`. It is modified as a side effect of calling `count`.
+
+Let's say we want to count fast, so we `jax.jit` the `count` method. (In this example, this wouldn't actually help speed anyway, for many reasons, but treat this as a toy model of wanting to jit-compile the update of model parameters, where `jax.jit` makes an enormous difference).
+
+```{code-cell} ipython3
+---
+id: 5jSjmJMon03W
+outputId: d952f16b-9b30-4753-ed94-cc914a929a36
+---
+counter.reset()
+fast_count = jax.jit(counter.count)
+
+for _ in range(3):
+  print(fast_count())
+```
+
++++ {"id": "weiI0V7_pKGv"}
+
+Oh no! Our counter isn't working. This is because the line
+```
+self.n += 1
+```
+in `count` was only called once, when JAX was compiling the method call. Moreover, since the return value doesn't depend on the arguments to `count`, once it returned the first 1, subsequent calls to `fast_count` will always return 1. This won't do. So, how do we fix it?
+
+## The solution: explicit state
+
+Part of the problem with our counter was that the returned value didn't depend on the arguments, meaning a constant was "baked into" the compiled output. But it shouldn't be a constant -- it should depend on the state. Well, then why don't we make the state into an argument?
+
+```{code-cell} ipython3
+---
+id: 53pSdK4KoOEZ
+outputId: 5ac72b9c-7029-4bf2-de8d-1d412bd74c79
+---
+from typing import Tuple
+
+CounterState = int
+
+class CounterV2:
+
+  def count(self, n: CounterState) -> Tuple[int, CounterState]:
+    # You could just return n+1, but here we separate its role as 
+    # the output and as the counter state for didactic purposes.
+    return n+1, n+1
+
+  def reset(self) -> CounterState:
+    return 0
+
+counter = CounterV2()
+state = counter.reset()
+
+for _ in range(3):
+  value, state = counter.count(state)
+  print(value)
+```
+
++++ {"id": "PrBjmgZtq89b"}
+
+In this new version of Counter, we moved `n` to be an argument of `count`, and added another return value that represents the new, updated, state. To use this counter, we now need to keep track of the state explicitly. But in return, we can now safely `jax.jit` this counter:
```

Comment:
Suggestion: nit (formatting)

```suggestion
In this new version of `Counter`, we moved `n` to be an argument of `count`, and added another return value that represents the new, updated, state. To use this counter, we now need to keep track of the state explicitly. But in return, we can now safely `jax.jit` this counter:
```

(similar to  "In our case, the `CounterV2` class is nothin..." further below.)